ledger/simulation/resources.go

```diff
@@ -0,0 +1,584 @@
+	// be populated if a top-level transaction executes AVM programs prior to v9 (when resource
+	// sharing was added).
+	localTxnResources []ResourceAssignment
+
+	startingBoxes int
+}
+
+func makeGroupResourceAssignment(txns []transactions.SignedTxnWithAD, proto *config.ConsensusParams) GroupResourceAssignment {
+	var startingBoxes int
+	var nonAppCalls int
+	localTxnResources := make([]ResourceAssignment, len(txns))
+	for i := range txns {
+		localTxnResources[i] = makeTxnResourceAssignment(&txns[i].Txn, proto)
+		startingBoxes += len(txns[i].Txn.Boxes)
+		if txns[i].Txn.Type != protocol.ApplicationCallTx {
+			nonAppCalls++
+		}
+	}
+	// Calculate the maximum number of cross-product resources that can be accessed by one app call
+	// under normal circumstances. This is calculated using the case of an app call with a full set
+	// of foreign apps. Including the app being called, there are (MaxAppTxnForeignApps + 1) apps,
+	// crossed with (MaxAppTxnForeignAssets + 2) accounts (the called app's account, the sender's
+	// account, and the foreign app accounts). We then subtract out the app local of sender's
+	// account and the called app, and each app local of an app and its own account, or
+	// (MaxAppTxnForeignApps + 2) references. So we end up with:
+	//
+	// (MaxAppTxnForeignApps + 1) * (MaxAppTxnForeignApps + 2) - (MaxAppTxnForeignApps + 2)
+	// <=> MaxAppTxnForeignApps^2 + 3*MaxAppTxnForeignApps + 2 - MaxAppTxnForeignApps - 2
+	// <=> MaxAppTxnForeignApps^2 + 2*MaxAppTxnForeignApps
+	// <=> MaxAppTxnForeignApps * (MaxAppTxnForeignApps + 2)
+	maxCrossProductsPerAppCall := proto.MaxAppTxnForeignApps * (proto.MaxAppTxnForeignApps + 2)
+	return GroupResourceAssignment{
+		Resources:                 makeGlobalResourceAssignment(localTxnResources, proto),
+		MaxCrossProductReferences: maxCrossProductsPerAppCall * (proto.MaxTxGroupSize - nonAppCalls),
+		localTxnResources:         localTxnResources,
+		startingBoxes:             startingBoxes,
+	}
+}
+
+func (a *GroupResourceAssignment) removePrivateFields() {
+	a.startingBoxes = 0
+	a.Resources.removePrivateFields()
+	for i := range a.localTxnResources {
+		a.localTxnResources[i].removePrivateFields()
+	}
+	a.localTxnResources = nil
+}
+
+func (a *GroupResourceAssignment) hasAccount(addr basics.Address, ep *logic.EvalParams, programVersion uint64, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		for i := range a.localTxnResources {
+			if a.localTxnResources[i].hasAccount(addr, ep, programVersion) {
+				return true
+			}
+		}
+		return a.Resources.hasAccount(addr, ep, programVersion)
+	}
+	return a.localTxnResources[txnIndex].hasAccount(addr, ep, programVersion)
+}
+
+func (a *GroupResourceAssignment) addAccount(addr basics.Address, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.addAccount(addr)
+	}
+	if !a.localTxnResources[txnIndex].addAccount(addr) {
+		return false
+	}
+	if a.Resources.hasAccount(addr, nil, 0) {
```

Comment:
Would you want to go so far as to remove the global cross-product ref if you demand both sides in some single transaction?

Reply:
No, I believe it's still useful to know that the actual cross-product is needed. With global resource sharing, even if the same txn references both sides of a cross-product, a possible assignment of those resources might place them in separate txns, so the cross-product wouldn't necessarily be available.

By specifying the cross-product, this serves as an additional constraint about where those two resources can be placed in the group.

Reply:
Are you placing resources into the local assignment, even if they can be satisfied by shared resources?  I thought they only go in if group sharing was insufficient, that is, `globalSharing` is false (which I assume happens for v < 9).

If that occurs for account A and asset Y in a single txn, then A and Y must go into the local txn (well, top-level ancestor of it), and AxY (if it exists) can come out of the global assignment.

I see my assumption about exactly when globalSharing is true is wrong - it's the point we discussed live one - but I think this comment still holds.

Reply:
Oh, I think I didn't fully understand the original question. You're asking about the situation where account A and asset Y happen to already both be in the local assignment for a txn, then a later txn that uses group sharing access the holding A x Y?

Right now we'd add A x Y to the global asset holdings, but we don't necessarily need to, since it's implicit in the earlier local assignment of A and Y to the same txn.

I have no issue with this in principal, but I'm worried there's a fair bit of complexity behind this corner, since if we do this, we'd probably also want to omit cross products where one half was referenced but unnamed and another half was named by the input txn (and for app calls, every named app also implies its account). Though if we want to implement #5616 in simulate, I guess we'd need code to check for this anyway...

Reply:
And it can happen in reverse. Currently, you remove new found local accounts from the global list. I was suggesting that you _might_ remove a holding from the global list if you found the account and the asset were both required locally. (in a later txn of the group that the original need for the holding that got it into the global list)

Reply:
Resolved during standup. Excluding the cross products described here would make for a more efficient/minimal response, but it's not necessary. Downstream code can still properly pack references in the txn group with this over-specification.

We can always revisit and decide to implement this in the future.